src/placement/mapless_jump_map.c

```diff
@@ -0,0 +1,1376 @@
+			setbit(dom_used, (selected_dom + child_pos));
+
+			top++;
+			curr_dom = &(curr_dom->do_children[selected_dom]);
+			obj_key = crc(obj_key, top);
+		}
+	} while (!found_target);
+
+}
+
+/**
+ * This function recursively chooses a single target to be used in the
+ * object shard layout. This function is called for every shard that needs a
+ * placement location.
+ *
+ * \param[in]   pmap		The pool map associated with this placement
+ *				map. This is used to directly access the
+ *				targets in the pool.
+ * \param[out]  target		Holds the value of the chosen spare target.
+ *				for the shard being rebuilt.
+ * \param[in]   key		A unique key generated using the object ID.
+ *				This is the same key used during initial
+ *				placement.
+ *				This is used in jump consistent hash.
+ * \param[in]   dom_used	This is a contiguous array that contains
+ *				information on whether or not an internal node
+ *				(non-target) in a domain has been used.
+ * \param[in]	layout		This is the current layout for the object.
+ *				This is needed for guaranteeing that we don't
+ *				reuse a target already in the layout.
+ * \param[in]	md		Object metadata used used to compare object
+ *				version with fail sequence.
+ *
+ * \return			Returns an error code if an error occurred,
+ *				otherwise 0.
+ */
+static int
+get_rebuild_target(struct pool_map *pmap, struct pool_target **target,
+		   uint64_t key, uint8_t *dom_used,
+		   struct pl_obj_layout *layout, struct daos_obj_md *md)
+{
+	uint8_t                 *used_tgts = NULL;
+	uint32_t                selected_dom;
+	uint32_t                fail_num = 0xFFc5;
+	uint32_t                try = 0;
+	struct pool_domain      *target_selection;
+	struct pool_domain      *root;
+	int rc = 0;
+
+	rc = pool_map_find_domain(pmap, PO_COMP_TP_ROOT, PO_COMP_ID_ALL, &root);
+	if (rc == 0) {
+		D_ERROR("Could not find root node in pool map.");
+		rc = -DER_NONEXIST;
+		return rc;
+	}
+
+	while (1) {
+
+		uint8_t skiped_targets = 0;
+		uint16_t num_bytes = 0;
+		uint32_t num_doms;
+
+		num_doms = root->do_child_nr;
+
+		uint64_t child_pos = (uint64_t)(root->do_children)
+				     - (uint64_t)root;
+		child_pos = child_pos / sizeof(struct pool_domain);
+
+		/*
+		 * If all of the nodes have been used for shards but we
+		 * still have shards to place mark all nodes as unused
+		 * so duplicates can be chosen
+		 */
+		if (isset_range(dom_used, child_pos, child_pos +
+				 num_doms - 1)) {
+			clrbit_range(dom_used, child_pos,
+				(child_pos + (num_doms - 1)));
+		}
+
+		/*
+		 * Choose domains using jump consistent hash until we find a
+		 * suitable domains that has not already been used.
+		 */
+		do {
+			key = crc(key, fail_num++);
+			selected_dom = jump_consistent_hash(key, num_doms);
+			target_selection = &(root->do_children[selected_dom]);
+		} while (isset(dom_used, (selected_dom + child_pos)));
+
+		/* Mark this domain as used */
+		setbit(dom_used, (selected_dom + child_pos));
```

Comment:
So rebuild target domain colocation problem isn't addressed? Will it be addressed in a follow-on patch?

Reply:
This should be addressed such that placement layouts will be as widely distributed as possible and if an object cannot be placed it will return an error before placement has started.